specifyweb/specify/record_merging.py

```diff
@@ -241,11 +245,35 @@ def record_merge_fx(model_name: str, old_model_ids: List[int], new_model_id: int
                         # EXAMPLE: ProtectedError: ("Cannot delete some instances of model 'Address' because they are 
                         # referenced through protected foreign keys:
                         # 'Division.address'.", {<Division: Division object (2)>})
-                        raise
+                        # 'Collectingevent.collectingeventattribute'", <QuerySet [<Collectingevent: Collectingevent object (2)>]>)
+                        match = re.search(r"'(\w+)\.(\w+)'", str(e))
+                        if match:
+                            protected_table, protected_field = match.groups()
+                            protected_model = getattr(spmodels, protected_table)
+                            # Check if this field is not required before updating it to None
+                            protected_dm_table = spmodels.datamodel.get_table(protected_table)
+                            protected_dm_field = protected_dm_table.get_field(protected_field) \
+                                if protected_dm_table else None
+                            if not protected_dm_table or not protected_dm_field or protected_dm_field.required:
+                                continue
+                            protected_model.objects.filter(**{protected_field: obj}).update(**{protected_field: None})
+                            try:
+                                obj.delete()
+                            except ProtectedError as e:
+                                raise
+                        else:
+                            raise
+
```

Comment:
> If we did, then we would have to pass collection and agent into the these function, which already has enough parameters.

that's the least of the problems in this case.

> unless you can think of a case where delete_resource works when the current code does not?

what if the dependent have dependents? then you'll have to recursively delete.
case: (Collectionobject -> CollectingEvent in embedded). so, when merging collection objects, your code will fail.

not saying your code doesn't handle it (partially, your code does handle one-hop away dependents -- nothing beyond that) but just that it is reinventing the other function. I think it is worth doing it now (well, if I was aware of this function initially, I'd have done it as part of https://github.com/specify/specify7/pull/3839 for the initial release). The minute we add more tables, we'll have to revisit this, again -- and we'll come-up with the exact rule -- delete dependents on deletion, and cascade it -- just what the function does -- already. 

The dependents belonging to other objects (the old ids) will get deleted (by that function) when the main object gets deleted. So why not delete it via that function? Otherwise, you've the awkward logic in the current code.

Did some refactoring (without the messy collection and agent argument pass you brought up). see https://github.com/specify/specify7/pull/4940/commits/a158853dca756924ebc40140d4531e3547df5b45. Moved tests to a separate directory + file (it was 1500 LOC in api_tests, no way). the backend tests pass without any change.

Reply:
See https://github.com/specify/specify7/pull/4940, and let me know if it can be merged to this PR.

Reply:
Ok, I see what you are saying now.  Ya, that looks good, I like how that simplifies the record merging code.  Go ahead and merge that branch into this PR